### PR TITLE
wal/wal.go: optimized WAL.SaveSnapshot to do Marshal outside the mutex lock

### PR DESCRIPTION
**Pull request by vimalk78**
modified :
- wal/wal.go
  modified WAL.SaveSnapshot  to do the Marshal before aquiring the mutex

### CODE REVIEW

**Comment by xiang90** [comment]
I do not think this is necessary. Most of WAL operations are serialized. I am fine with simply moving marshal out of the protection of lock though. It makes the code looks more correct.

/cc @heyitsanthony

**Comment by vimalk78** [comment]
@xiang90 thats what i have done, move the Marshal outside the lock. do you see any other change in WAL? can you pls point out the same, i will revert it.
i will also revert the wal_bench_test.go also
/cc @heyitsanthony

File: wal/wal.go [hunk 1]
```diff
@@ -568,10 +583,12 @@ func (w *WAL) Save(st raftpb.HardState, ents []raftpb.Entry) error {
 }
 
 func (w *WAL) SaveSnapshot(e walpb.Snapshot) error {
+
```
Comment (xiang90) [inline wal/wal.go]: remove the extra empty line at line 586

File: wal/wal_bench_test.go [hunk 2]
```diff
@@ -49,13 +49,15 @@ func benchmarkWriteEntry(b *testing.B, size int, batch int) {
 	for i := 0; i < size; i++ {
 		data[i] = byte(i)
 	}
-	e := &raftpb.Entry{Data: data}
+	e := raftpb.Entry{Data: data}
```
Comment (xiang90) [inline wal/wal_bench_test.go]: revert these changes.

File: wal/wal.go [hunk 3]
```diff
@@ -532,21 +540,28 @@ func (w *WAL) saveState(s *raftpb.HardState) error {
 }
 
 func (w *WAL) Save(st raftpb.HardState, ents []raftpb.Entry) error {
-	w.mu.Lock()
```
Comment (xiang90) [inline wal/wal.go]: revert the changes in this func? let's first improve save entry by saving one allocation? then we can move the encoding out of lock?

Comment (vimalk78) [inline wal/wal.go]: saving one allocation will require having a static byte buffer and using the same buffer with invoking MarshalTo(data []byte) ? 

if we use a static byte buffer, we will need the mutex lock always. right?

Comment (xiang90) [inline wal/wal.go]: we can use sync.Pool

File: wal/wal.go [hunk 4]
```diff
@@ -521,6 +521,14 @@ func (w *WAL) saveEntry(e *raftpb.Entry) error {
 	return nil
 }
 
+func (w *WAL) saveEntryOpt(data []byte) error {
+	rec := &walpb.Record{Type: entryType, Data: data}
+	if err := w.encoder.encode(rec); err != nil {
```
Comment (heyitsanthony) [inline wal/wal.go]: `return w.encoder.encode(rec)`?

File: wal/wal.go [hunk 5]
```diff
@@ -521,6 +521,14 @@ func (w *WAL) saveEntry(e *raftpb.Entry) error {
 	return nil
 }
 
+func (w *WAL) saveEntryOpt(data []byte) error {
```
Comment (heyitsanthony) [inline wal/wal.go]: call this `saveEntry` and get rid of the old `saveEntry`? Probably not much of a reason to have both `saveEntryOpt` and `saveEntry` functions.

File: wal/wal.go (hunk 3 above)
Comment (vimalk78) [inline wal/wal.go]: some thoughts:
 - protobuf compiler generated method "func (m *Entry) MarshalTo(data []byte) (int, error)" needs a buffer sized Entry.Size() or more. this method will not append to the buffer
 - the way fmt package uses sync.Pool cannot be used here. fmt creates a handle which holds the buffer (initially unallocated) and uses append to grow the buffer. the generated code in raft.pb.go does not use this approach
 - we could use a simple approach of using sync.Pool with initial size of 1024. if this is not sufficient, we can allocate a new buf, use it and push to the Pool. i have added a snippet here : https://play.golang.org/p/fLKjslWKd6
 - saving this allocation is not enough, as "func (e *encoder) encode(rec *walpb.Record)" is going to marshal the Record, which is going to allocate again. only difference is Record will also contain the type. it is better we allocate for the Record and use MarshalTo() using the slices in it for Data/MetaData/Snapshot/Crc etc.

Comment (vimalk78) [inline wal/wal.go]: @xiang90 @heyitsanthony any comments

Comment (heyitsanthony) [inline wal/wal.go]: I'd rather avoid complicating the code with sync.Pool unless there's a meaningful performance impact when running the etcd server.

Comment (xiang90) [inline wal/wal.go]: I agree with @heyitsanthony. Let's just revert this change for now? And keep the change for SaveSnapshot?

Comment (vimalk78) [inline wal/wal.go]: committed the changes for SaveSnapshot . can you check once.

Comment (vimalk78) [inline wal/wal.go]: shall i close this?

**Comment by heyitsanthony** [comment]
lgtm
